packages/x-internals/src/forwardRefShim/forwardRefShim.tsx

```diff
@@ -0,0 +1,16 @@
+import * as React from 'react';
+import reactMajor from '../reactMajor';
+
+// Compatibility shim that ensures stable props object for forwardRef components
+// Fixes https://github.com/facebook/react/issues/31613
+// We ensure that the ref is always present in the props object (even if that's not the case for older versions of React) to avoid the footgun of spreading props over the ref in the newer versions of React.
+// Footgun: <Component ref={ref} {...props} /> will break past React 19, but the types will now warn us that we should use <Component {...props} ref={ref} /> instead.
+export const forwardRefShim = <T, P = {}>(
+  render: React.ForwardRefRenderFunction<T, P & { ref: React.Ref<T> }>,
+) => {
+  if (reactMajor >= 19) {
+    const Component = (props: any) => render(props, props.ref ?? null);
+    return Component as React.ForwardRefExoticComponent<P>;
```

Comment:
Might be nice to carry the `render.name` or `render.displayName`, it's useful when profiling performance.

Reply:
Agree, I think displayName only should be enough – added.